insonmnia/worker/overseer.go

```diff
@@ -657,3 +663,111 @@ func (o *overseer) OnDealFinish(ctx context.Context, containerID string) error {
 func (o *overseer) Logs(ctx context.Context, id string, opts types.ContainerLogsOptions) (io.ReadCloser, error) {
 	return o.client.ContainerLogs(ctx, id, opts)
 }
+
+func (o *overseer) removeImageRepoAndTag(image io.Reader) (io.Reader, error) {
+	decompressed, err := o.extract(image)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decompress image: %v", err)
+	}
+
+	var pruned []tarHeader2file
+	for _, h2f := range decompressed {
+		switch h2f.header.Name {
+		case "manifest.json":
+			prunedManifest, err := o.getPrunedImageManifest(h2f.file.Bytes())
+			if err != nil {
+				return nil, fmt.Errorf("failed to prune image manifest: %v", err)
+			}
+			h2f.file = bytes.NewBuffer(prunedManifest)
+			h2f.header.Size = int64(len(prunedManifest))
+		}
+		pruned = append(pruned, h2f)
+	}
+
+	compressed, err := o.compress(pruned)
+	if err != nil {
+		return nil, fmt.Errorf("failed to re-compress image: %v", err)
+	}
+
+	return compressed, nil
+}
+
+func (o *overseer) extract(r io.Reader) ([]tarHeader2file, error) {
+	var tr = tar.NewReader(r)
+	var out []tarHeader2file
+	for {
+		header, err := tr.Next()
+		if err == io.EOF {
+			return out, nil
+		}
+
+		if err != nil {
+			return nil, fmt.Errorf("failed to read header: %v", err)
+		}
+
+		if header == nil {
+			continue
+		}
+
+		switch header.Typeflag {
+		case tar.TypeDir:
+			out = append(out, tarHeader2file{header: header, file: nil})
+		case tar.TypeReg:
+			var buf = bytes.NewBuffer([]byte{})
+			if _, err := io.Copy(buf, tr); err != nil {
+				return nil, err
+			}
+			out = append(out, tarHeader2file{header: header, file: buf})
+		}
+	}
+
+	return nil, errors.New("unexpected exit")
+}
+
+func (o *overseer) compress(decompressed []tarHeader2file) (io.Reader, error) {
+	buf := bytes.NewBuffer([]byte{})
```

Comment:
it's quite expensive in terms of memory usage and can fail with oom. Please consider buffering on disk.